### PR TITLE
CRv2: Process roles attribute

### PR DESCRIPTION
**Pull request by hacodeorg**
[PLC-918](https://codedotorg.atlassian.net/browse/PLC-918): Use contact's `roles` to set Pardot prospect's `db_Roles`. 
* Contact's `roles` value is combined from multiple sources
    - dashboard.census_submissions
    - dashboard.census_submissions#submitter_role
    - dashboard.pd_enrollments
    - dashboard.pd_attendances
    - dashboard.followers
    - dashboard.sections#course_name
    - dashboard.sections#curriculum_umbrella and course_name
    - dashboard.user_permissions#permission
    - dashboard.users#is_parent
    - dashboard.users#user_id
    - pegasus.forms
    - pegasus.forms#kind
* Prospect's [db_Roles](https://pi.pardot.com/prospectFieldCustom/read/id/6743) is a multi-select attribute.

## Links
[Speadsheet with all attributes to process](https://docs.google.com/spreadsheets/d/10oKPBVlC5LZee9WiECwd745sX-4EXFALuSIVVwdW5tc/edit#gid=0).

## Test story
Test data extraction and processing on an adhoc with connection to a production-cloned db.
* Test the updated `ContactRollupsRaw.extract_scripts_taught` method
    ```ruby
    def run_extraction
      puts "Truncating ContactRollupsRaw...";
      ActiveRecord::Base.connection.truncate(ContactRollupsRaw.table_name)
      puts "ContactRollupsRaw.count before = #{ContactRollupsRaw.count}"
      puts "Running the given block..."
      yield
      puts "ContactRollupsRaw.count after = #{ContactRollupsRaw.count}"
      puts "First 3 rows:";
      puts ContactRollupsRaw.first(3).pluck(:email, :sources, :data, :data_updated_at);
      puts "Last 3 rows:";
      puts ContactRollupsRaw.last(3).pluck(:email, :sources, :data, :data_updated_at)
      nil
    end

    # Verify data->'$.curriculum_umbrella' in ContactRollupsRaw records
    run_extraction {ContactRollupsRaw.extract_scripts_taught}
    ```
 
* Test the entire extraction and procesing
    ```ruby
    @limit = 100
    cr = ContactRollupsV2.new(is_dry_run: true, limit_extraction: @limit);

    # clean up existing tables
    cr.truncate_or_delete_table ContactRollupsRaw
    cr.truncate_or_delete_table ContactRollupsProcessed

    # roles attribute uses data from 10 out of 14 extraction methods, 
    # so just run them all
    cr.collect_contacts

    # dry run. Verify the sample queries sent to Pardot
    cr.process_contacts
    cr.sync_updated_contacts_with_pardot
    cr.sync_new_contacts_with_pardot
    cr.report_results

    # Verify roles values in ContactRollupsProcessed
    sql_cr_processed = <<-SQL.squish
      select data->'$.roles' as db_Roles, count(*) as hit
      from contact_rollups_processed
      group by db_Roles order by hit desc
    SQL
    ActiveRecord::Base.connection.exec_query(sql_cr_processed)

    # Verify db_Roles values in ContactRollupsPardotMemory
    sql_cr_pm = <<-SQL.squish
      select data_synced->'$.db_Roles' as db_Roles, count(*) as hit
      from (select * from contact_rollups_pardot_memory limit 10000) as sq
      group by db_Roles order by hit desc
    SQL
    ActiveRecord::Base.connection.exec_query(sql_cr_pm)
    ```

### CODE REVIEW

File: dashboard/app/models/contact_rollups_processed.rb [hunk 1]
```diff
@@ -183,6 +207,49 @@ def self.extract_professional_learning_attended(contact_data)
     return uniq_courses.blank? ? {} : {professional_learning_attended: uniq_courses}
   end
 
+  def self.extract_roles(contact_data)
+    roles = []
+    # Contact is a teacher if they appears in any of the following tables
+    roles << 'Teacher' if
+      contact_data.dig('dashboard.users', 'user_id') ||
```
Comment (molly-moen) [inline dashboard/app/models/contact_rollups_processed.rb:221]: do only teachers have a `user_id`?

File: dashboard/app/models/contact_rollups_processed.rb [hunk 2]
```diff
@@ -183,6 +207,49 @@ def self.extract_professional_learning_attended(contact_data)
...
+      contact_data.key?('dashboard.pd_enrollments') ||
+      contact_data.key?('dashboard.pd_attendances') ||
+      contact_data.key?('dashboard.followers')
+
+    # Contact is a teacher if they submit a census survey as a teacher
+    submitter_roles = extract_field(contact_data, 'dashboard.census_submissions', 'submitter_role') || []
```
Comment (molly-moen) [inline dashboard/app/models/contact_rollups_processed.rb]: do we need to do this check if we already know the user is a teacher?

File: dashboard/app/models/contact_rollups_processed.rb (hunk 1 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_processed.rb:221]: At this stage of the pipeline, yes. In the previous data collection step, we select only teacher accounts (only teacher accounts have emails). The teacher filer is  on line 129.
https://github.com/code-dot-org/code-dot-org/blob/7a99855caef4dcf160c178e7b00405980952af13/dashboard/app/models/contact_rollups_raw.rb#L124-L132

File: dashboard/app/models/contact_rollups_processed.rb (hunk 2 above)
Comment (hacodeorg) [reply dashboard/app/models/contact_rollups_processed.rb]: That's a good point. I will add a check to avoid redundant work.

**Review by molly-moen — APPROVED** [review]

**Pull request merged**